File: go/bob/bob.go
// Package bob responds according to your input.
package bob

import (
	"regexp"
	"strings"
)

// Hey forwards remark to bob and returns bob's answer.
func Hey(remark string) string {
	r := strings.Trim(remark, " \t\n\r")
	if r == "" {
		return "Fine. Be that way!"
	}
	if IsShoutingQuestion(r) {
		return "Calm down, I know what I'm doing!"
	}
	if IsShouting(r) {
		return "Whoa, chill out!"
	}
	if IsQuestion(r) {
		return "Sure."
	}
	return "Whatever."
}

// IsQuestion checks whether remark is a question.
func IsQuestion(remark string) bool {
	s, _ := regexp.MatchString("\\?$", remark)
	return s
}

// IsShouting checks whether remark is UPPERCASE and does not contain any lower case.
func IsShouting(remark string) bool {
	s, _ := regexp.MatchString("[A-Z]{2,}", remark)
	r, _ := regexp.MatchString("[a-z]", remark)
	return s && (r == false)
}

func IsShoutingQuestion(remark string) bool {
	return IsShouting(remark) && IsQuestion(remark)
}

File: elixir/bob/bob.exs
defmodule Bob do

    def hey(string) do
        cond do
            just_whitespace?(string) -> "Fine. Be that way!"
            is_all_upper?(string) && is_question?(string) -> "Calm down, I know what I'm doing!"
            is_all_upper?(string) -> "Whoa, chill out!"
            is_question?(string) -> "Sure."
            true -> "Whatever."
        end
    end

    def remove_number(string) do
        :binary.list_to_bin(
            String.split(string, ~r{\d}, trim: true)
        )
    end

    def just_whitespace?(string) do
        Enum.all?(
            String.to_charlist(string),
            fn char -> [char] == ' ' end
        )
    end

    def is_all_upper?(string) do
        upper = String.upcase(string)
        string == upper and String.match?(remove_number(string), ~r/[[:alpha:]]/)
    end

    def is_question?(string) do
        "?" == String.last(string)
    end

end